Subject: Various mail fixes
Last-Update: 2010-04-11

* rDoS fix: Unchecked contents of MAILTO would allow sending arbitrary commands
  to sendmail. Prevent this.
* Don't terminate when finding a dot "." on a line by itself. (sendmail -i)
* Don't force immediate delivery
* Style/formatting

Bug-Debian: http://bugs.debian.org/36338
Bug-Debian: http://bugs.debian.org/146224


POSSIBLE BUG:
  
   mailto += strspn(mailto, " \t\n");

why is mailto altered, apparently without reversing the effect?

The entire MAILTO sanitation code appears a bit dodgy and could use a rewrite
(ie, a correct parsing for a valid local or remote address).

NOT-Acked as of 4010-04-11
Index: patched/do_command.c
===================================================================
--- patched.orig/do_command.c	2010-05-06 18:18:30.896560600 +0200
+++ patched/do_command.c	2010-05-06 18:18:31.634435779 +0200
@@ -96,6 +96,21 @@
 	usernm = env_get("LOGNAME", e->envp);
 	mailto = env_get("MAILTO", e->envp);
 
+	/* Check for arguments */
+	if (mailto) {
+		const char	*end;
+
+		/* These chars have to match those cron_popen()
+		 * uses to split the command string */
+		mailto += strspn(mailto, " \t\n");
+		end = mailto + strcspn(mailto, " \t\n");
+		if (*mailto == '-' || *end != '\0') {
+			printf("Bad Mailto karma.\n");
+			log_it("CRON",getpid(),"error","bad mailto");
+			mailto = NULL;
+		}
+	}
+
 #ifdef USE_SIGCHLD
 	/* our parent is watching for our death by catching SIGCHLD.  we
 	 * do not care to watch for our children's deaths this way -- we
Index: patched/config.h
===================================================================
--- patched.orig/config.h	2010-05-06 18:18:14.096435742 +0200
+++ patched/config.h	2010-05-06 18:18:31.635435778 +0200
@@ -42,22 +42,27 @@
 			 */
 
 #define MAILCMD _PATH_SENDMAIL					/*-*/
-#define MAILARGS "%s -FCronDaemon -odi -oem -or0s %s"		/*-*/
-			/* -Fx	 = set full-name of sender
+/* #define MAILARGS "%s -i -FCronDaemon -odi -oem  %s"		/*-*/
+#define MAILARGS "%s -i -FCronDaemon -oem  %s"		/*-*/
+			/* -i    = don't terminate on "." by itself
+                         * -Fx	 = set full-name of sender
 			 * -odi	 = Option Deliverymode Interactive
 			 * -oem	 = Option Errors Mailedtosender
+ 			 * -t    = read recipient from header of message
 			 * -or0s = Option Readtimeout -- don't time out
+			 * XXX: sendmail doesn't allow -or0s when invoked
+			 * by joe user.  --okir
 			 */
 
-/* #define MAILCMD "/bin/mail"			/*-*/
-/* #define MAILARGS "%s -d  %s"			/*-*/
+/* #define MAILCMD "/bin/mail"			-*/
+/* #define MAILARGS "%s -d  %s"			-*/
 			/* -d = undocumented but common flag: deliver locally?
 			 */
 
-/* #define MAILCMD "/usr/mmdf/bin/submit"	/*-*/
-/* #define MAILARGS "%s -mlrxto %s"		/*-*/
+/* #define MAILCMD "/usr/mmdf/bin/submit"	-*/
+/* #define MAILARGS "%s -mlrxto %s"		-*/
 
-/* #define MAIL_DATE				/*-*/
+/* #define MAIL_DATE				-*/
 			/* should we include an ersatz Date: header in
 			 * generated mail?  if you are using sendmail
 			 * for MAILCMD, it is better to let sendmail
@@ -68,7 +73,7 @@
 			 * defined but neither exists, should crontab(1) be
 			 * usable only by root?
 			 */
-/*#define ALLOW_ONLY_ROOT			/*-*/
+/*#define ALLOW_ONLY_ROOT			-*/
 
 			/* if you want to use syslog(3) instead of appending
 			 * to CRONDIR/LOG_FILE (/var/cron/log, e.g.), define
